_posts/2023-04-04-s3-sns-notification.markdown

```diff
@@ -0,0 +1,120 @@
+
+## How to Avoid "Unable to validate the following destination configurations" Error
+
+As is widely known, working with CloudFormation as an Infrastructure as Code tool to deploy resources poses a major challenge when it comes to deciphering errors in the create/update stack process. These errors are not usually straightforward or informative, which makes it difficult to identify the problem immediately. Recently, I had a good fun with reconfigure notifications on an Amazon S3 bucket, and it took me a full day to trouble shoot why an error was occurring and how to resolve it. I consulted numerous resources, each of which pointed to a specific reason for the deployment failure. Ultimately, I had to combine all the information I had gathered to successfully address the issue.
+
+This brief blog post will outline the complexities involved in updating SNS Notification configurations on an Amazon S3 Bucket, as well as steps for resolving the issue efficiently.
+
+## Why "Invalid Destination" Happens?
+
+The reason for this is largely due to the way Cloudformation orders the dependencies. The S3 Notification configurations are considered part of the Amazon S3 Bucket properties and are created simultaneously with the bucket. To ensure proper resource creation, Cloudformation must follow this sequence:
+
+- Create the SNS Topic.
+- Create the Amazon S3 Bucket.
+- Create the SNS Access Policy to allow your S3 Bucket to publish events to the specified topic.
+- Create the S3 Notification Configuration.
+
+Now, let's examine the Cloudformation code for deploying the Amazon S3 Bucket and SNS Topics.
+
+```
+----- main.yml ------
+AWSTemplateFormatVersion: 2010-09-09
+Description: SNS topic and S3 Notification configuration
+
+Parameters:
+  MySNSTopicName:
+    Type: String
+  MyBucketName:
+    Type: String
+
+Resources:
+  NewTopic:
+    Type: AWS::SNS::Topic
+    Properties:
+      TopicName: !Ref MySNSTopicName
+
+  DataBucket:
+    Type: AWS::S3::Bucket
+    DeletionPolicy: Retain
+    DependsOn: !Ref Topic
+    UpdateReplacePolicy: Retain
+    Properties:
+      BucketName: !Ref MyBucketName
+      VersioningConfiguration:
+        Status: Enabled
+
+  TopicPolicy:
+    Type: AWS::SNS::TopicPolicy
+    Properties:
+      PolicyDocument:
+        Id: BucketNotificationPolicy
+        Version: "2012-10-17"
+        Statement:
+          - Sid: AllowPublishNotifications
+            Effect: Allow
+            Principal:
+              Service: "s3.amazonaws.com"
+            Action: sns:Publish
+            Resource:
+              - !Ref NewTopic
+            Condition:
+              ArnLike:
+                aws:SourceArn: !GetAtt DataBucket.Arn
+```
+
+Once you have deployed the SNS Topic, S3 Bucket, and SNS Access Policy, the next step is to configure S3 Notifications for the specified SNS topic. This can be accomplished easily with the following block of code:
```

Comment:
`Once you have deployed the SNS Topic, S3 Bucket, and SNS Access Policy`

TO:
`Once you have deployed the SNS Topic, S3 Bucket and SNS Access Policy` (grammar)